core/translate/main_loop.rs

```diff
@@ -480,6 +472,131 @@ pub fn init_loop(
+    program: &mut ProgramBuilder,
+    t_ctx: &mut TranslateCtx,
+    table_references: &TableReferences,
+    predicates: &[WhereTerm],
+    hash_join_op: &HashJoinOp,
+    mode: OperationMode,
+) -> Result<()> {
+    // Resolve cursors for the build table
+    let build_table = &table_references.joined_tables()[hash_join_op.build_table_idx];
+    let (table_cursor_id, _index_cursor_id) = build_table.resolve_cursors(program, mode)?;
+
+    // If the build table doesn't currently have a table cursor (e.g., plan chose a
+    // covering index and only opened the index cursor), allocate and open a table
+    // cursor so the hash join can later SeekRowid into the base table if needed.
+    let table_cursor_id = if let Some(cursor_id) = table_cursor_id {
+        cursor_id
+    } else {
+        // Allocate a cursor for the build table
+        let cursor_id = program.alloc_cursor_id_keyed_if_not_exists(
+            CursorKey::table(build_table.internal_id),
+            match &build_table.table {
+                Table::BTree(btree) => CursorType::BTreeTable(btree.clone()),
+                _ => panic!("Hash join build table must be a BTree table"),
+            },
+        );
+        if let Table::BTree(btree) = &build_table.table {
+            program.emit_insn(Insn::OpenRead {
+                cursor_id,
+                root_page: btree.root_page,
+                db: build_table.database_id,
+            });
+        }
+        cursor_id
+    };
+    let iteration_cursor_id = table_cursor_id;
+
+    // use the build table's InternalTableID for the hash table reference
+    let hash_table_reg: usize = build_table.internal_id.into();
+    let num_keys = hash_join_op.join_keys.len();
+    let build_key_start_reg = program.alloc_registers(num_keys);
+
+    // Create new loop for hash table build phase
+    let build_loop_start = program.allocate_label();
+    let build_loop_end = program.allocate_label();
+    let skip_to_next = program.allocate_label();
+
+    program.emit_insn(Insn::Rewind {
+        cursor_id: iteration_cursor_id,
+        pc_if_empty: build_loop_end,
+    });
+
+    program.preassign_label_to_next_insn(build_loop_start);
+
+    // Planner invariants for the hash-build side:
+    // - join_lhs_and_rhs only chooses a HashJoin access method when the
+    //   left-hand (build) table's access method is a plain table scan:
+    //   That means the build table is scanned via its base B-tree and is
+    //   not driven by any index or search constraints.
+    // - Because the build side is a full table scan, all build-side join
+    //   key expressions can safely read their columns directly from the
+    //   table cursor. We never need DeferredSeek in the build phase.
+    //
+    // We intentionally do NOT apply WHERE clause filters during hash build.
+    // The hash build phase runs independently of the main join loop, so conditions
+    // that reference other tables (which aren't open yet) would fail.
```

Comment:
shouldn't we be able to detect this? although if there's some independent WHERE filter on the build side then maybe hash join is a poor planning decision for that case. but this seems like bit of a hack to me